Added statistics 1

Moliholy opened the pull request:
Here I have modified some files with the new statistics system. There are more to be modified, but I think I have enough changes to take a look at them so that the next ones are done according to what is discussed here.

jblomer commented:
Alright, looking through it.  Please have a look at cpplint errors first.

jblomer commented:
I think it would make sense to make the variable names consistent with the counter name.  So for instance the `certificate_hits_` counter variable would become `n_certificate_hits_`.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2105,15 +2104,24 @@ static int Init(const loader::LoaderExports *loader_exports) {
+      "Number of opened files");
+  cvmfs::num_fs_dir_open_ = cvmfs::statistics_->Register("cvmfs.no_fs_dir_open",
+      "Number of opened directories");
+  cvmfs::num_fs_lookup_ = cvmfs::statistics_->Register("cvmfs.no_fs_loopup",
```
That would be `n_fs_lookup`, same for the other counters below (ever growing counters)

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2626,7 +2637,7 @@ static int AltProcessFlavor(int argc, char **argv) {
   }
   if (strcmp(argv[1], "__wpad__") == 0) {
-    return download::MainResolveProxyDescription(argc, argv);
+    return download::MainResolveProxyDescription(argc, argv, cvmfs::statistics_);
```
Here we should create the statistics class in MainResolveProxyDescription.  This is one of the "special modes" in which the client can be started.  In this particular special mode, it doesn't act as a fuse module but only resolves the proxy definition and prints the result.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2811,8 +2822,8 @@ static bool RestoreState(const int fd_progress,
       SendMsg2Socket(fd_progress, "Restoring open files counter... ");
-      cvmfs::open_files_ = *(reinterpret_cast<uint32_t *>(
-        saved_states[i]->state));
+      cvmfs::open_files_->Set(*(reinterpret_cast<uint64_t *>(
```
Ah, here we have a slightly bigger problem... let's discuss.

Moliholy commented:
Added the rest of the files to the new statistics system

Moliholy commented on cvmfs/download.cc:
```diff
@@ -422,8 +422,8 @@ void *DownloadManager::MainDownload(void *data) {
-      download_mgr->statistics_->transfer_time +=
-        DiffTimeSeconds(timeval_start, timeval_stop);
+      perf::Xadd(download_mgr->counters_->sz_transfer_time,
+        1000 * DiffTimeSeconds(timeval_start, timeval_stop));
```
Now we are counting miliseconds, not seconds. This is because now the counter is not a double, but a long, so in order be more accurate I had to change it

Moliholy commented on cvmfs/download.cc:
```diff
@@ -1516,10 +1517,10 @@ Failures DownloadManager::Fetch(JobInfo *info) {
     do {
       retval = curl_easy_perform(handle);
-      statistics_->num_requests++;
+      perf::Inc(counters_->n_num_requests);
```
In principle this operation was not atomic, but now it is

Moliholy commented on cvmfs/download.cc:
```diff
@@ -1516,10 +1517,10 @@ Failures DownloadManager::Fetch(JobInfo *info) {
       double elapsed;
       if (curl_easy_getinfo(handle, CURLINFO_TOTAL_TIME, &elapsed) == CURLE_OK)
-        statistics_->transfer_time += elapsed;
+        perf::Xadd(counters_->sz_transfer_time, (int64_t)(elapsed * 1000));
```
Mandatory cast of a double into a int64_t to get the miliseconds

Moliholy commented:
Sorry, I forgot to add the catalog_mgr_*** statistics. Now it is fixed.

jblomer commented:
Before every pull request, please make a cpplint run

jblomer commented:
The t_lru.cc unit test needs to be adapted to the new LruCache interface.

Moliholy commented:
Fixed talk.cc to not to print old statistics.
Fixed all unittests, making them compatible with the new statistics. I have just run the 475 tests and they all successfully passed.
I have also tried the program out (only basic operations, nothing big) and I found no crashes.

jblomer commented on cvmfs/cache.h:
```diff
@@ -80,8 +82,8 @@ class CatalogManager : public catalog::AbstractCatalogManager {
     return result;
   }
   std::string GetCertificateStats() {
```
This function can be removed I guess.

Moliholy replied:
Removed, even though it's not shown here

jblomer replied:
When I try to merge, this function is still present.  Can you have a look?

jblomer commented on cvmfs/catalog_mgr.h:
```diff
@@ -67,38 +68,37 @@ inline const char *Code2Ascii(const LoadError error) {
+    n_lookup_xattrs = statistics->Register("catalog_mgr.n_lookup_xattrs",
+        "Number of xattrs lookups");
+    n_listing = statistics->Register("catalog_mgr.n_listing",
+        "Number of listing");
```
"Number of listing" --> "Number of listings"

jblomer commented on cvmfs/catalog_mgr.h:
```diff
@@ -67,38 +68,37 @@ inline const char *Code2Ascii(const LoadError error) {
+    n_listing = statistics->Register("catalog_mgr.n_listing",
+        "Number of listing");
+    n_nested_listing = statistics->Register("catalog_mgr.n_nested_listing",
+        "Number of nested listing");
```
"Number of nested listing" --> Number of listings of nested catalogs

jblomer commented on cvmfs/catalog_mgr.h:
```diff
@@ -67,38 +68,37 @@ inline const char *Code2Ascii(const LoadError error) {
+        "Number of nested listing");
   }
 
   std::string Print() {
```
This function can be removed

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Statistics *statistics_ = NULL;
+perf::Counter *no_fs_open_ = NULL;
+perf::Counter *no_fs_dir_open_ = NULL;
+perf::Counter *no_fs_lookup_ = NULL;
```
`no_fs_lookup` --> `n_fs_lookup`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_open_ = NULL;
+perf::Counter *no_fs_dir_open_ = NULL;
+perf::Counter *no_fs_lookup_ = NULL;
+perf::Counter *no_fs_lookup_negative_ = NULL;
```
`no_fs_lookup_negative_` --> `n_fs_lookup_negative_`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_dir_open_ = NULL;
+perf::Counter *no_fs_lookup_ = NULL;
+perf::Counter *no_fs_lookup_negative_ = NULL;
+perf::Counter *no_fs_stat_ = NULL;
```
`no_fs_stat_` --> `n_fs_stat_`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_lookup_ = NULL;
+perf::Counter *no_fs_lookup_negative_ = NULL;
+perf::Counter *no_fs_stat_ = NULL;
+perf::Counter *no_fs_read_ = NULL;
```
`no_fs_read_` --> `n_fs_read`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_lookup_negative_ = NULL;
+perf::Counter *no_fs_stat_ = NULL;
+perf::Counter *no_fs_read_ = NULL;
+perf::Counter *no_fs_readlink_ = NULL;
```
`no_fs_readlink_` --> `n_fs_readlink`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_stat_ = NULL;
+perf::Counter *no_fs_read_ = NULL;
+perf::Counter *no_fs_readlink_ = NULL;
+perf::Counter *no_fs_forget_ = NULL;
```
`no_fs_forget_` --> `n_fs_forget_`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *no_fs_read_ = NULL;
+perf::Counter *no_fs_readlink_ = NULL;
+perf::Counter *no_fs_forget_ = NULL;
+perf::Counter *no_io_error_ = NULL;
```
`no_io_error_` --> n_io_error_

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
-atomic_int32 open_dirs_; /**< number of currently open directories */
+perf::Statistics *statistics_ = NULL;
+perf::Counter *no_fs_open_ = NULL;
+perf::Counter *no_fs_dir_open_ = NULL;
```
`no_fs_dir_open_` --> `n_fs_dir_open_`

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
-atomic_int32 open_files_; /**< number of currently open files by Fuse calls */
-atomic_int32 open_dirs_; /**< number of currently open directories */
+perf::Statistics *statistics_ = NULL;
+perf::Counter *no_fs_open_ = NULL;
```
`no_fs_open_` --> `n_fs_open_`

Moliholy commented:
All current fixes applied

jblomer commented on cvmfs/catalog_mgr.h:
```diff
@@ -67,38 +68,27 @@ inline const char *Code2Ascii(const LoadError error) {
+    n_listing = statistics->Register("catalog_mgr.n_listing",
+        "Number of listings");
+    n_nested_listing = statistics->Register("catalog_mgr.n_nested_listing",
+        "Number of nested listings");
```
Description: number of listings of nested catalogs

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *n_fs_readlink_ = NULL;
+perf::Counter *n_fs_forget_ = NULL;
+perf::Counter *n_io_error_ = NULL;
+/**< number of currently open files by Fuse calls */
```
I think the Doxygen `/**<` syntax only works for comments that are in the same line.  For comments in front of a code line, the

    /**
     * comment
    */

syntax should be used.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -222,18 +222,20 @@ uint64_t next_directory_handle_ = 0;
+perf::Counter *n_io_error_ = NULL;
+/**< number of currently open files by Fuse calls */
+perf::Counter *no_open_files_ = NULL;
+/**< number of currently open directories */
```
Same as above

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -339,12 +341,12 @@ static bool UseWatchdog() {
 
-void GetLruStatistics(lru::Statistics *inode_stats, lru::Statistics *path_stats,
-                      lru::Statistics *md5path_stats)
+void GetLruStatistics(lru::Counters *inode_stats, lru::Counters *path_stats,
```
This function can be removed

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -368,20 +370,15 @@ catalog::Statistics GetCatalogStatistics() {
-  return catalog_manager_->GetCertificateStats();
-}
-
 string GetFsStats() {
```
This function can be removed

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1070,7 +1067,7 @@ static void cvmfs_open(fuse_req_t req, fuse_ino_t ino,
   }
 
-  atomic_inc64(&num_fs_open_);  // Count actual open / fetch operations
+  perf::Inc(n_fs_open_);;  // Count actual open / fetch operations
```
double semicolon

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1520,11 +1517,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
     attribute_value = StringifyInt(max_open_files_ - kNumReservedFd);
   } else if (attr == "user.usedfd") {
-    attribute_value = StringifyInt(atomic_read32(&open_files_));
+    attribute_value = no_open_files_->Print();
```
At the moment, `Print()` does the right thing.  However, `Print()` in cvmfs is used for human-readable output, so the exact format of the string might change in the future.  Please add a function `ToString()` to the `perf::Counter` class, which transforms the number into a string.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1520,11 +1517,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
+    attribute_value = no_open_files_->Print();
   } else if (attr == "user.useddirp") {
-    attribute_value = StringifyInt(atomic_read32(&open_dirs_));
+    attribute_value = no_open_dirs_->Print();
```
See above.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1520,11 +1517,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
+    attribute_value = no_open_dirs_->Print();
   } else if (attr == "user.nioerr") {
-    attribute_value = StringifyInt(atomic_read32(&num_io_error_));
+    attribute_value = n_io_error_->Print();
```
See above.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1566,9 +1563,9 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
     attribute_value = StringifyInt(num_catalogs);
   } else if (attr == "user.nopen") {
-    attribute_value = StringifyInt(atomic_read64(&num_fs_open_));
+    attribute_value = n_fs_open_->Print();
```
See above

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1566,9 +1563,9 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
+    attribute_value = n_fs_open_->Print();
   } else if (attr == "user.ndiropen") {
-    attribute_value = StringifyInt(atomic_read64(&num_fs_dir_open_));
+    attribute_value = n_fs_dir_open_->Print();
```
See above

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1580,11 +1577,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
     attribute_value = StringifyInt(seconds_direct);
   } else if (attr == "user.rx") {
-    int64_t rx = uint64_t(download_manager_->GetStatistics().transferred_bytes);
+    int64_t rx = download_manager_->GetCounters().no_transferred_bytes->Get();
```
This can be replaced by `Statistics->Lookup("n_transferred_bytes")`, in order to get rid of the `GetCounters()` method.

jblomer replied:
Hm, on a second thought: it is probably better to leave it in.  This way, the compiler will complain whenever we change the name of the counter in the download manager.  Never mind.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1580,11 +1577,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
   } else if (attr == "user.speed") {
-    int64_t rx = uint64_t(download_manager_->GetStatistics().transferred_bytes);
-    int64_t time = uint64_t(download_manager_->GetStatistics().transfer_time);
+    int64_t rx = download_manager_->GetCounters().no_transferred_bytes->Get();
```
See above.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1580,11 +1577,11 @@ static void cvmfs_getxattr(fuse_req_t req, fuse_ino_t ino, const char *name,
-    int64_t rx = uint64_t(download_manager_->GetStatistics().transferred_bytes);
-    int64_t time = uint64_t(download_manager_->GetStatistics().transfer_time);
+    int64_t rx = download_manager_->GetCounters().no_transferred_bytes->Get();
+    int64_t time = download_manager_->GetCounters().sz_transfer_time->Get();
```
See above.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2105,15 +2110,24 @@ static int Init(const loader::LoaderExports *loader_exports) {
-  atomic_init64(&cvmfs::num_fs_forget_);
-  atomic_init32(&cvmfs::num_io_error_);
+  cvmfs::n_fs_open_ = cvmfs::statistics_->Register("cvmfs.n_fs_open",
+      "Number of opened files");
```
Description: Overall number of file open operations

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2105,15 +2110,24 @@ static int Init(const loader::LoaderExports *loader_exports) {
+  cvmfs::n_fs_open_ = cvmfs::statistics_->Register("cvmfs.n_fs_open",
+      "Number of opened files");
+  cvmfs::n_fs_dir_open_ = cvmfs::statistics_->Register("cvmfs.n_fs_dir_open",
+      "Number of opened directories");
```
Description: Overall number of directory open operations

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2105,15 +2110,24 @@ static int Init(const loader::LoaderExports *loader_exports) {
+  cvmfs::n_fs_readlink_ = cvmfs::statistics_->Register("cvmfs.n_fs_readlink",
+      "Number of links read");
+  cvmfs::n_fs_forget_ = cvmfs::statistics_->Register("cvmfs.n_fs_forget",
+      "Number of forgets");
```
Description: number of inode forgets

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2281,8 +2295,10 @@ static int Init(const loader::LoaderExports *loader_exports) {
-  atomic_init32(&cvmfs::open_files_);
-  atomic_init32(&cvmfs::open_dirs_);
+  cvmfs::no_open_files_ = cvmfs::statistics_->Register("cvmfs.no_open_files",
+      "Total number of opened files");
```
Description: Number of currently open files

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2281,8 +2295,10 @@ static int Init(const loader::LoaderExports *loader_exports) {
+  cvmfs::no_open_files_ = cvmfs::statistics_->Register("cvmfs.no_open_files",
+      "Total number of opened files");
+  cvmfs::no_open_dirs_ = cvmfs::statistics_->Register("cvmfs.no_open_dirs",
+      "Total number of opened directories");
```
Description: number of currently open directories

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -2626,7 +2644,8 @@ static int AltProcessFlavor(int argc, char **argv) {
   if (strcmp(argv[1], "__wpad__") == 0) {
-    return download::MainResolveProxyDescription(argc, argv);
+    return download::MainResolveProxyDescription(argc, argv,
+        cvmfs::statistics_);
```
This is a special mode of operation.  The Statistics object should be created on the fly in `download::MainResolveProxyDescription()`

jblomer commented on cvmfs/cvmfs.h:
```diff
@@ -46,8 +46,8 @@ std::string GetOpenCatalogs();
 void ResetErrorCounters();
-void GetLruStatistics(lru::Statistics *inode_stats, lru::Statistics *path_stats,
-                      lru::Statistics *md5path_stats);
+void GetLruStatistics(lru::Counters *inode_stats, lru::Counters *path_stats,
```
Not needed anymore

jblomer commented on cvmfs/download.h:
```diff
@@ -83,25 +84,31 @@ inline const char *Code2Ascii(const Failures error) {
+struct Counters {
+  perf::Counter *no_transferred_bytes;
+  perf::Counter *sz_transfer_time;  // measured in miliseconds
+  perf::Counter *n_num_requests;
```
`n_num_requests` --> `n_requests`

jblomer commented on cvmfs/download.h:
```diff
@@ -83,25 +84,31 @@ inline const char *Code2Ascii(const Failures error) {
+  perf::Counter *no_transferred_bytes;
+  perf::Counter *sz_transfer_time;  // measured in miliseconds
+  perf::Counter *n_num_requests;
+  perf::Counter *n_num_retries;
```
`n_num_retries` --> `n_retries`

jblomer commented on cvmfs/download.h:
```diff
@@ -83,25 +84,31 @@ inline const char *Code2Ascii(const Failures error) {
-    num_proxy_failover = 0;
-    num_host_failover = 0;
+struct Counters {
+  perf::Counter *no_transferred_bytes;
```
`no_transferred_bytes` --> `sz_transferred`

jblomer commented on cvmfs/download.h:
```diff
@@ -83,25 +84,31 @@ inline const char *Code2Ascii(const Failures error) {
+        "Number of host failovers");
   }
 
   std::string Print() const;
```
Shouldn't be required anymore

jblomer commented on cvmfs/download.h:
```diff
@@ -301,7 +309,7 @@ class DownloadManager {
   void GetTimeout(unsigned *seconds_proxy, unsigned *seconds_direct);
   void SetLowSpeedLimit(const unsigned low_speed_limit);
-  const Statistics &GetStatistics();
+  const Counters &GetCounters();
```
Not needed anymore

jblomer commented on cvmfs/lru.h:
```diff
@@ -71,49 +72,56 @@ namespace lru {
-  atomic_int64 num_insert;
-  atomic_int64 num_insert_negative;
+struct Counters {
+  perf::Counter *size;
```
`size` --> `sz_size`

jblomer commented on cvmfs/lru.h:
```diff
@@ -71,49 +72,56 @@ namespace lru {
+  perf::Counter *sz_allocated;
+
+  Counters(perf::Statistics *statistics, const std::string &name) {
+    size = statistics->Register(name + ".no_size", "Size for " + name);
```
Name: `no_size` --> `sz_size`

jblomer commented on cvmfs/lru.h:
```diff
@@ -71,49 +72,56 @@ namespace lru {
+        "Number of allocated bytes for " + name);
   }
 
   std::string Print() {
```
Not needed anymore

jblomer commented on cvmfs/nfs_shared_maps.cc:
```diff
@@ -206,20 +208,17 @@ bool GetPath(const uint64_t inode, PathString *path) {
 }
 
 
 string GetStatistics() {
```
Not needed anymore

Moliholy replied:
It is used in the line 259 of the nfs_maps.cc file:

```
string GetStatistics() {
  if (use_shared_db_)
    return nfs_shared_maps::GetStatistics();

  string result = "Total number of issued inodes: " +
                  StringifyInt(seq_-root_inode_) + "\n";

  string stats;
  db_inode2path_->GetProperty(leveldb::Slice("leveldb.stats"), &stats);
  result += "inode --> path database:\n" + stats + "\n";

  db_path2inode_->GetProperty(leveldb::Slice("leveldb.stats"), &stats);
  result += "path --> inode database:\n" + stats + "\n";

  return result;
}
```

Shall we remove that if (used_shared_db_) ?

jblomer replied:
Not only the if but the whole `GetStatistics()` in nfs_maps.cc can go away, no?

Moliholy replied:
It is used in swissknife_pull.cc, line 558. However, it's just a log that can be perfectly removed

jblomer replied:
This comment seems to be misplaced (is for `DownloadManager::GetStatistics()`, right?).  Please see my follow-up comment.  It probably makes sense to keep the method in `DownloadManager`.

jblomer commented on cvmfs/shortstring.h:
```diff
@@ -175,15 +176,34 @@ class ShortString {
-  static uint64_t num_instances() { return atomic_read64(&num_instances_); }
-  static uint64_t num_overflows() { return atomic_read64(&num_overflows_); }
+  static void RegisterCounters(perf::Statistics *statistics) {
+    if (StackSize == kDefaultMaxPath && Type == 0) {
```
Checking on `Type` should be sufficient

jblomer commented on cvmfs/shortstring.h:
```diff
@@ -175,15 +176,34 @@ class ShortString {
 
-  static uint64_t num_instances() { return atomic_read64(&num_instances_); }
-  static uint64_t num_overflows() { return atomic_read64(&num_overflows_); }
+  static void RegisterCounters(perf::Statistics *statistics) {
```
Hm, that's not too nice yet.  If one forgets to use `RegisterCounters`, it will result in a crash.  I'll have to think about it.

jblomer replied:
Let's try like this: add a function `perf::TryInc` that will only increase the counter if it is not `NULL`.  This new function should be used instead of `perf::Inc` in the `ShortString` class (but only there).

Moliholy replied:
Fixed in the last update

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
+
 namespace perf {
 
+Counter::Counter() { atomic_init64(&counter_); }
```
This needs to go back to the header, otherwise it can't be inlined.

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
 namespace perf {
 
+Counter::Counter() { atomic_init64(&counter_); }
+void Counter::Inc() { atomic_inc64(&counter_); }
```
See above.

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
 
+Counter::Counter() { atomic_init64(&counter_); }
+void Counter::Inc() { atomic_inc64(&counter_); }
+void Counter::Dec() { atomic_dec64(&counter_); }
```
See above

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
+Counter::Counter() { atomic_init64(&counter_); }
+void Counter::Inc() { atomic_inc64(&counter_); }
+void Counter::Dec() { atomic_dec64(&counter_); }
+int64_t Counter::Get() { return atomic_read64(&counter_); }
```
See above

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
+void Counter::Inc() { atomic_inc64(&counter_); }
+void Counter::Dec() { atomic_dec64(&counter_); }
+int64_t Counter::Get() { return atomic_read64(&counter_); }
+void Counter::Set(const int64_t val) { atomic_write64(&counter_, val); }
```
See above

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -7,12 +7,40 @@
+void Counter::Dec() { atomic_dec64(&counter_); }
+int64_t Counter::Get() { return atomic_read64(&counter_); }
+void Counter::Set(const int64_t val) { atomic_write64(&counter_, val); }
+int64_t Counter::Xadd(const int64_t delta) {
```
See above

jblomer commented on cvmfs/statistics.cc:
```diff
@@ -40,7 +68,7 @@ string Statistics::PrintList(const PrintOptions print_options) {
        iEnd = counters_.end(); i != iEnd; ++i)
   {
-    result += i->first + "|" + StringifyInt(i->second->counter.Get()) +
+    result += i->first + "|" + i->second->counter.Print() +
```
Better use the `ToString()` method (see above)

jblomer commented on cvmfs/wpad.cc:
```diff
@@ -226,7 +226,8 @@ static void AltCvmfsLogger(const LogSource source, const int mask,
 
-int MainResolveProxyDescription(int argc, char **argv) {
+int MainResolveProxyDescription(int argc, char **argv,
+    perf::Statistics *statistics) {
```
Here you can create the `Statistics` object on the fly (see above)

jblomer commented on mount/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ set (CVMFS_MOUNT_SOURCES
   ../cvmfs/logging.h ../cvmfs/logging_internal.h ../cvmfs/logging.cc
   ../cvmfs/options.h ../cvmfs/options.cc
   ../cvmfs/sanitizer.h ../cvmfs/sanitizer.cc
+  ../cvmfs/statistics.h ../cvmfs/statistics.cc
```
Is this really needed?

Moliholy replied:
Actually no, it isn't. My mistake. Now it is fixed

jblomer commented:
Hm, something is still not quite right.  If I try to mount, I get the error "/usr/lib64/libcvmfs_fuse.so: undefined symbol: _ZN8nfs_maps13GetStatisticsEv".  Can you confirm?

Also, please always use cpplint.py before submitting the pull request.

Moliholy commented:
Yes, there was a mistake. I called a function defined in the .h that was recently deleted by mistake in the corresponding .cc
After the rebase it should be fixed (for me it works now), as well as the cpplint.

jblomer commented on cvmfs/cvmfs.h:
```diff
@@ -46,8 +46,6 @@ std::string GetOpenCatalogs();
-                      lru::Statistics *md5path_stats);
 std::string PrintInodeTrackerStatistics();
 std::string PrintInodeGeneration();
 catalog::Statistics GetCatalogStatistics();
```
This declaration should be removed (I think there is at least one more that got removed in the .cc file but not here)

Moliholy commented:
Fixed the unnecessary method in cache.h

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -339,15 +345,6 @@ static bool UseWatchdog() {
-}
-
-
 string PrintInodeTrackerStatistics() {
```
This function and similar ones can be removed, too.

Moliholy replied:
Removed as well

jblomer commented:
There is one bigger problem: running the unit tests results in a memory corruption.  Can you have a look?

Moliholy commented:
I have just run the unittests three times without apparently any problem

Moliholy commented:
So, I have fixed the bug! It was a problem with the initialization of the counters, apparently. This is because the "statistics objects" are created depending on the binary in different places, so it's quite difficult to keep track of all of them. I would suggest to create a _super_ global singleton object in statistics.h so that every class who wants to register its counters just have to import this file and call the object method.

reneme commented:
I've been working with @Moliholy on this today and I really wonder why the central `Statistics` class isn't designed as a singleton for three reasons:

1. The statistics class is a cross-cutting concern and therefore should be accessible from essentially everywhere.
2. A singleton inherently does its memory management automagically via lazy initialisation even before `main()` runs, if necessary.
3. It would simplify the API of using classes be entirely decoupling the `Statistics` class from them.
4. Naïvely, I don't see a reason for more than one `Statistics` object per process.

Especially point (2) is interesting, because it would allow us to get rid of the `ShortString<>::RegisterCounter()` methods that initialise global counter state in a rather icky way. Furthermore, the global `Statistics` object doesn't need to be handled by some high-level program code anymore.

I just wanted to cross-check if there is a reason to not make it a singleton.

jblomer commented:
You need multiple statistics objects for the different contexts in the library.  Otherwise there is a name clash.  Other than that I have no objection against singleton.

reneme commented:
Statistics are per repository?

jblomer commented:
Every counter has a name.  That would clash.

reneme commented:
Yes, I got that... What about prefixing this name with the repo's FQRN?

jblomer commented:
Fine with me, if you find a way to figure out the fqrn at early initialization time of the statistics object.

Moliholy commented:
I have a expected problem with some counters. When executing the following in libcvmfs_test:

```
/home/molina/cvmfs/build/cvmfs/test_libcvmfs: initializing with options: cache_directory=/tmp/test-libcvmfs-cache
cvmfs> attach sft
attaching repo with options: repo_name=sft.cern.ch,url=http://cvmfs-stratum-one.cern.ch/opt/sft;http://cernvmfs.gridpp.rl.ac.uk/opt/sft;http://cvmfs.racf.bnl.gov/opt/sft,pubkey=/etc/cvmfs/keys/cern.ch.pub
CernVM-FS: using public key(s) /etc/cvmfs/keys/cern.ch.pub
cvmfs> attach atlas
attaching repo with options: repo_name=atlas.cern.ch,url=http://cvmfs-stratum-one.cern.ch/opt/atlas;http://cernvmfs.gridpp.rl.ac.uk/opt/atlas;http://cvmfs.racf.bnl.gov/opt/atlas,pubkey=/etc/cvmfs/keys/cern.ch.pub
test_libcvmfs: /home/molina/cvmfs/cvmfs/statistics.cc:72: perf::Counter* perf::Statistics::Register(const std::string&, const std::string&): Assertion `counters_.find(name) == counters_.end()' failed.
```

There is (obviously) a name clash when registering a counter. To fix it there are two quick options:

1) Identify the classes that are invoking these methods and force them to add a "surname" (scope) to the variable. In a nutshell, these classes would be the followings:
  * cache::CatalogManager
  *  download::DownloadManager
  *  lru::Md5PathCache
However, this would make the functionality for these classes more complicated and it would add more characters to the counter name, which is maybe not desirable.

2) A even quicker and simpler workaround would be to remove the restriction in the Counter::Register method so that it now returns the instance that has already been created. This way these "numbers" would be modified by two different context, but in any case at the moment this is only possible in libcvmfs, whose statistics are not shown. For a future pull request this could (should) probably be modified adding a different solution.


What's your opinion?

jblomer commented:
The second option looks appealing but it is also dangerous if at any point we use the same name for different counters by mistake.  I think there is also another issue with the ShortString class and the hotpatching.  In fact it would be better not to touch the memory layout of this class, otherwise we need to code the memory layout upgrade.  I'd suggest we have a little white board session next week, now that we know where all the dead bodies are hidden.

jblomer commented:
As discussed, to be replaced by two smaller pull requests.  Closing this one.
